### PR TITLE
Multiple requests

### PR DESCRIPTION
**Pull request by noa-neria**
# Concurrent multi-request streaming with capacity-controlled object storage

## Goal

Support **many concurrent read submissions** and let them saturate object-storage bandwidth
as well as a single large submission does. Previously a submission was drained before the next
one ran; now a new `_ex` API accepts many submissions at once and each worker **interleaves**
them, keeping every client's in-flight window full regardless of how the reads were split across
submissions.

## Changes

- **Multi-request API** — `runai_request` / `runai_response` (C, plus Python bindings in
  `libstreamer.py`). Submit returns a **submission id**; responses carry the owning submission id
  and a `submission_done` flag. Many submissions may be in flight simultaneously over one
  persistent responder, demuxed by submission id.  The legacy `runai_request` / `runai_response`
  remain unchanged, and `file_streamer` still uses them. `runai_response_ex` accepts timeout in ms
  
  Note: legacy `runai_request` and `runai_response`  changed to the new behavior. Python layer modified accordingly.

- **`ObjectStorageWorker`** — a per-thread async worker (`cpp/streamer/impl/object_storage_worker/`)
  built on a reusable `CapacityWorker` base (`cpp/utils/capacity_worker/`) and a credit-bounded
  `CapacityQueue`. One worker keeps its client's in-flight window full across **several** workloads
  at once, so a small/late submission is served alongside a large one. Total throughput stays
  `s3_concurrency × per-client window`.

- **`BackendPools`** — one thread pool per backend kind: synchronous filesystem reads and
  asynchronous object-storage workers, created lazily on first use.

- **`Workload`** reduced to a plain validated batch container; async scheduling moved into the
  worker, and the unused per-task global-id system removed.

- **`SubmissionsMgr`** — per-submission bookkeeping (id allocation, remaining-response counting,
  per-submission throughput) behind its own leaf mutex.

- **Window initialization & errors** — `capacity()` builds the client and sizes the in-flight
  window, throwing if it cannot (missing plugin, empty workload). The base funnels the throw to a
  single `discard` hook that finalizes the workload with a meaningful code (e.g. `S3NotSupported`)
  and retries the window on the next workload.

## How the user detects the end of a submission

Completion is **per submission**, reported through the response API — there is no separate
"wait for submission" call.

1. `runai_request_ex` returns a **submission id** and records how many responses that submission
   will produce (one per requested sub-range).

2. The user consumes responses in a loop with `runai_response_ex`. Every successful response
   returns `(submission_id, file_index, index, submission_done)`:
   - `submission_id` tells the user **which** submission the response belongs to (responses from
     different in-flight submissions are interleaved and may arrive in any order).
   - `submission_done` is `true` **only on that submission's last response** — i.e. the streamer
     has now delivered every sub-range of that submission.

   Internally the streamer decrements a per-submission remaining-response counter as each response
   is consumed; when it reaches zero the record is retired and that final response is flagged
   `submission_done = true` (and the submission's throughput is logged).

3. To know a specific submission has fully arrived, the user watches for the response whose
   `submission_id` matches and whose `submission_done` is `true`. To drain several submissions,
   the user keeps consuming until it has seen a `submission_done` for each id it submitted.

Notes:
- Because the responder is **persistent**, a drained-but-open streamer does **not** report
  `FinishedError`; `runai_response_ex` blocks (or times out) waiting for more work.
  `FinishedError` is returned only on teardown.
- `runai_response_ex(timeout_ms)` returns a timeout indication if no response arrives in time
  (surfaced as `TimeoutError` in Python), so a consumer can poll without blocking forever.

## Summary

- **New Features**
  - Added extended multi-file streaming APIs (`runai_request_ex` / `runai_response_ex`) that demultiplex by `submission_id` and report when a submission is fully finished.
  - Added `runai_set_credentials` for streamer-scoped object-storage credentials, with enforced single-backend/single-credential usage per streamer.
  - Improved timed response draining with clearer timeout handling for drained vs persistent behavior.
- **Bug Fixes**
  - Improved concurrent submission correctness and stop/drain edge-case reliability.
- **Tests**
  - Expanded C++ and Python coverage for multi-submission demux, timeouts, credential propagation, and teardown stress.

### CODE REVIEW

File: cpp/streamer/impl/submissions/submissions_mgr.cc [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    {
+        const auto & sub = it->second;
+        const auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(_now() - sub.submit_time).count();
+
+        result.outcome = Result::Outcome::Completed;
+        result.total_bytes = sub.total_bytes;
+        result.elapsed_ms = ms;
+        result.throughput_bps = (ms > 0) ? (sub.total_bytes * 1000 / static_cast<size_t>(ms)) : 0;
+
+        _submissions.erase(it);
```
[the remark this answers is not in the document]
Comment (noa-neria) [reply cpp/streamer/impl/submissions/submissions_mgr.cc]: This project's ASSERT is a custom macro (cpp/utils/logging/logging.h:137), not C assert(), and it is not stripped in release.

File: cpp/streamer/impl/object_storage_worker/object_storage_worker.h [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+// so the base owns the in-flight window (a CapacityQueue of ObjectChunk) and runs the submit/drain
+// interleave; this class only fills the backend hooks. Because the base interleaves, ONE worker
+// can hold chunks from SEVERAL workloads in flight at once, refilling the window from whichever workload
+// has pending chunks - so a small/late submission is served without waiting for a large one to drain.
+//
+// It is plugin-agnostic (hence the name, not S3Worker): it drives a Reader over an S3ClientWrapper, which
+// dispatches to the s3/gcs/azure plugin by URI. The reader/client is built once, on the first workload,
+// and reused for every later workload on this thread (client reuse is keyed on credentials, not bucket,
+// so multi-bucket submissions share it; a credentials change is not yet handled - single-credential first
+// cut, TODO drain-and-switch).
```
[the remark this answers is not in the document]
Comment (noa-neria) [reply cpp/streamer/impl/object_storage_worker/object_storage_worker.h]: Streamer now verifies that credentials have not changed, and return error otherwise. Common use case the default credential chain, and not by passing credentials explicetly

File: cpp/streamer/impl/pools/backend_pools.cc [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        if (ready != static_cast<int>(plugin))
+        {
+            LOG(ERROR) << "Streamer is locked to a single object storage plugin; rejecting a submission using a different plugin";
+            return common::ResponseCode::UnsupportedBackendMix;
+        }
+        if (credentials.empty())
+        {
+            return common::ResponseCode::Success;
+        }
+        // non-empty credentials must be verified under the mutex (fall through)
```
[the remark this answers is not in the document]
Comment (noa-neria) [reply cpp/streamer/impl/pools/backend_pools.cc]: Credentials are now set once for the entire streamer scope

File: cpp/streamer/streamer.cc [hunk 4]
```diff
@@ -82,12 +138,7 @@ _RUNAI_EXTERN_C int runai_request(
...
     void ** dsts,
     unsigned * num_sizes,
-    size_t ** internal_sizes,
-    const char * key,
-    const char * secret,
-    const char * token,
-    const char * region,
-    const char * endpoint
+    size_t ** internal_sizes
 )
```
[the remark this answers is not in the document]
Comment (noa-neria) [reply cpp/streamer/streamer.cc:142]: The only users of the C API are in the streamer pyton artifacts, which are shipped together with the corresponding cpp library and with the same version

File: py/runai_model_streamer/runai_model_streamer/file_streamer/file_streamer.py [hunk 5]
```diff
@@ -80,54 +81,56 @@ def __exit__(self, exc_type: any, exc_value: any, traceback: any) -> None:
...
+        # First object-storage path only: resolve the credentials and apply them to the streamer, once. boto3
+        # resolution is AWS-only, so it runs only for S3 paths (is_s3_path) - the backend type is only known
+        # from the path; the resolved credentials are then applied by the C++ layer to the plugin the URI
+        # selects. Nothing is applied when there is nothing to apply (no credentials -> the C++ default/ambient
+        # provider chain is used, and the streamer's set-once credentials stay free).
+        if s3_credentials_module and is_s3_path(path) and self.s3_session is None:
+            self.s3_session, self.s3_credentials = s3_credentials_module.get_credentials(credentials)
+            if streamer is not None and self._has_credentials():
+                runai_set_credentials(streamer, self.s3_credentials)
         return path
```
[the remark this answers is not in the document]
Comment (noa-neria) [reply py/runai_model_streamer/runai_model_streamer/file_streamer/file_streamer.py:103]: But `__enter__ `resets it (file_streamer.py:70-71):
```
def __enter__(self):
    self.streamer = runai_start()
    ...
    self.s3_session = None       # <-- resolution flag cleared
    self.s3_credentials = None
```
So entering the context makes the guard `self.s3_session` is None fire `True` again → creds are re-resolved and applied to T2.

The coupling between session creation and applying credentials should be resolved in the CR fix

File: cpp/utils/capacity_worker/capacity_worker.h [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+                drain_batch(stopped);                   // window full -> free slots, then submit more
+            }
+        }
+        if (_queue->inflight() >= _queue->capacity())
+        {
+            drain_batch(stopped);                       // fully submitted but full -> free a slot for the next
+        }
+    }
+
+    void drain(std::atomic<bool> & stopped) final override
```
[the remark this answers is not in the document]
Comment (noa-neria) [reply cpp/utils/capacity_worker/capacity_worker.h:70]: Valid hang — confirmed. Note the proposed `discard(std::move(request))` in `execute()` wouldn't fix the common case: enqueue takes `Workload&&` by reference and the move into `_inflight` (object_storage_worker.cc:157) happens before the throwing allocations (`resize/push_back/queue enqueue`), so request is already moved-from and discard would push no responses. It also wouldn't clear chunks already enqueued into the window (→ submit()'s unknown-handle ASSERT).
Fixed inside `enqueue` with a try/catch that calls `abort_all(UnknownError)`. OOM (like any UnknownError) means the caller aborts the streamer, so we don't reconstruct exact per-response counts — `abort_all` fails this worker's in-flight workloads (this one included, since the realistic OOM point is the bulk allocation after the move, so it's already in `_inflight`) and zeroes the window, clearing the half-built entry and any pre-throw queued chunks. The one corner it doesn't individually finalize is a throw in the emplace itself (pre-move) — a single small node allocation whose failure means the process is already out of memory and being torn down, so best-effort there is fine

File: py/runai_model_streamer/runai_model_streamer/libstreamer/libstreamer.py [hunk 7]
```diff
@@ -91,18 +154,97 @@ def runai_response(streamer: t_streamer) -> Optional[Tuple[int, int]]:
...
+    if error_code == FINISHED_ERROR_CODE:
+        return None
+    if error_code == TIMED_OUT_ERROR_CODE:
+        raise TimeoutError(
+            f"runai_response_ex timed out after {timeout_ms} ms"
+        )
+    if error_code != SUCCESS_ERROR_CODE:
+        raise ValueError(
+            f"Could not receive runai_response_ex from libstreamer due to: {runai_response_str(error_code)}"
+        )
```
[the remark this answers is not in the document]
Comment (noa-neria) [reply py/runai_model_streamer/runai_model_streamer/libstreamer/libstreamer.py]: **Resolved** — the finding was correct for the commit reviewed, but the code has since changed.

`runai_response` (renamed from `runai_response_ex`) no longer raises on a per-sub-range error. It returns the full tuple for every real response:

```
if error_code == FINISHED_ERROR_CODE:
    return None
if error_code == TIMED_OUT_ERROR_CODE:
    raise TimeoutError(...)
return (error_code, submission_id.value, file_index.value, range_index.value, bool(submission_done.value))
```

So a per-range error that is the submission's last response comes back as `(error_code, sid, file_index, index, submission_done=True)` — the submission_done the C layer set is preserved, and a multi-submission consumer's if done: `pending.discard(sid)` works without orphaning the ID. Errors are surfaced as data; the caller decides policy (abort only on `UnknownError`, or fail just that submission and keep draining the rest).

The only remaining raise-on-error is in` FileStreamer.request_ready_chunks`, which is a single-submission fail-fast consumer — it has no sibling submission to orphan, so the hang described here doesn't apply.

Test coverage added for this exact scenario:

- `test_libstreamer.py::test_response_error_is_returned_not_raised` — an EOF error that is the submission's last response; asserts runai_response returns a tuple with a non-Success response_code, the matching `submission_id`, and `submission_done=True` (not raised, not None). The filesystem mock's error path was also made faithful to the real C API (it now sets the out-params + submission_done on an error response).
- `test_file_streamer.py::test_get_chunks_raises_on_read_error` — asserts `FileStreamer` still fails fast (raises `ValueError`) on that error.

File: cpp/streamer/impl/streamer/streamer.cc [hunk 8]
```diff
@@ -151,36 +212,96 @@ common::ResponseCode Streamer::async_request(
...
+            // route to the pool for this workload's backend kind (a workload is homogeneous)
+            const auto kind = workloads[next].is_object_storage()
+                ? BackendPools::Kind::ObjectStorage
+                : BackendPools::Kind::FileSystem;
+            _pools.push(kind, std::move(workloads[next]));
         }
     }
 
-    __responder_release.cancel();
+    drain_guard.cancel(); // all workloads dispatched
```
[the remark this answers is not in the document]
Comment (noa-neria) [reply cpp/streamer/impl/streamer/streamer.cc:283]: **Resolved** — reviewed against an older commit. The premise ("`out_submission_id` is not written on the error path") no longer holds.

`out_submission_id` is now written the instant the id is minted, *before* any dispatch — `streamer.cc:208-212`:

    const unsigned submission_id = _submissions.generate();
    if (out_submission_id != nullptr)
    {
        *out_submission_id = submission_id;
    }

So on a dispatch throw after commit (the OOM path where `drain_undispatched` fires), the caller receives **both** the error code **and** the submission id. Those `total_sizes` responses are still delivered (dispatched workloads complete normally; undispatched ones are drained as `UnknownError`), and the caller can consume them by that id — they are not "ghost responses carrying an unknown submission_id."

This is intentional and documented at `streamer.cc:200-207`: the id is reported once it exists regardless of how the call ends — on success it identifies the submission, and on a post-commit failure it lets the caller drain the submission's already-accounted responses. (A *pre-commit* failure returns the error with `out_submission_id` left 0 and nothing committed — no increment, no responses — so there's nothing to drain and nothing to orphan.)

Matches the caller contract: on `UnknownError` the caller typically aborts and tears the streamer down (`~Streamer` stops the responder, unblocking any consumer).

File: cpp/streamer/impl/object_storage_worker/object_storage_worker.cc [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+void ObjectStorageWorker::abort_all(common::ResponseCode code)
+{
+    for (auto it = _inflight.begin(); it != _inflight.end(); )
+    {
+        auto next = std::next(it);
+        finalize(it, code);   // fails every batch and erases `it`
+        it = next;
+    }
+
+    // zero the window so idle() becomes true and the pool can join: move any pending chunk into flight,
```
[the remark this answers is not in the document]
Comment (noa-neria) [reply cpp/streamer/impl/object_storage_worker/object_storage_worker.cc]: **Declining — the re-entry it guards against no longer happens, so the guard would be dead code.**

This reviewed an older commit. In the current `drain_batch`, a stale/late backend completion does **not** route to `abort_all`. When `locate()` returns `_inflight.end()`, the completion is dropped and scanning continues (`object_storage_worker.cc:388-398`):

    auto [wlit, task_idx] = locate(response.handle);
    if (wlit == _inflight.end())
    {
        // ... Never abort_all here: that would fail every OTHER in-flight submission over a stray event.
        LOG(DEBUG) << "Dropping late object storage completion with unknown handle " << response.handle;
        continue;
    }

So `abort_all` is never re-entered on a stale completion — that path is explicitly a drop.

The only in-round `abort_all` (line 408) is gated on `!progressed && responder_drained && _queue != nullptr && !_queue->idle()`. After the first `abort_all` zeros the window, `_queue->idle()` is true, so `!_queue->idle()` is false — it can't fire again. `abort_all` is idempotent regardless (the `_inflight` loop over an empty map is a no-op and the window is already zeroed), but with the drop + the `!idle()` gate it isn't reached a second time in the first place.

Adding `if (_inflight.empty()) return;` would therefore never execute in a scenario that matters, and the invariant it hints at is already documented in the `abort_all` header comment (late completions are dropped; `abort_all` runs only on a terminal client state or the OOM-in-`enqueue` path). Prefer to leave it out.

Reasonable to reject this one — the team's style has leaned toward documenting invariants over adding defensive guards that can't execute.

File: py/runai_model_streamer/runai_model_streamer/file_streamer/file_streamer.py [hunk 10]
```diff
@@ -176,25 +189,33 @@ def get_chunks(self) -> Iterator:
...
     # The indexes are relative to the last request that sent
     # And need to be translated to global index in the chunks list
     def request_ready_chunks(self) -> Iterator:
+        # Only one submission is in flight at a time (get_chunks fully drains the active request before
+        # submitting the next), so every response here belongs to self.submission_id and the count loop is
+        # exact. runai_response blocks indefinitely (timeout 0) and returns None only on teardown.
         for i in range(sum(len(file_request.chunks) for file_request in self.active_request.files)):
-            file_relative_index, chunk_relative_index = runai_response(self.streamer)
-            if chunk_relative_index == None:
+            response = runai_response(self.streamer)
```
[the remark this answers is not in the document]
Comment (noa-neria) [reply py/runai_model_streamer/runai_model_streamer/file_streamer/file_streamer.py:207]: **Valid finding, but the suggested fix (a `busy()`-style guard) is the opposite of where this is headed — so we're deferring the real fix rather than re-adding the guard.**

Two parts:

**1. Not reachable for current users.** The shipped consumers (vLLM, sglang) never reuse a failed streamer — on a `ValueError` out of `get_chunks` they let it propagate and tear the streamer down. So the corruption isn't hit in practice today.

**2. The correct fix is submission_id demux, not a busy-guard.** `FilesRequestsIteratorWithBuffer` currently allocates one large buffer and reuses it, which forces each submission to be fully drained before the buffer is reused — hence today's single-submission, count-bounded `request_ready_chunks`. The next branch replaces that with a **ring buffer** so buffers are reused without a full drain, which means `FileStreamer` will manage **concurrent** submissions. Re-adding a `busy()` guard would block exactly that.

The multi-request API was built for this: every response already carries its `submission_id` and `submission_done` (including error responses — see the earlier thread). Once the ring buffer lands, `request_ready_chunks` will route each response to its owning submission by `submission_id` (and drain/finalize per submission), so a failed submission's leftover responses are attributed to *it*, never misfed to another stream. That removes the silent-corruption path structurally.

So for this branch we're leaving `FileStreamer` single-submission (its actual contract today) and doing the demux with the ring-buffer change, rather than adding a guard we'd immediately remove. An assert of the expected `submission_id` will throw on corruption.

File: cpp/streamer/impl/pools/backend_pools.cc [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+    {
+        // First object-storage submission: record the plugin and create the pool, then publish readiness.
+        // _ready_plugin is stored AFTER the pool exists (release), so a fast-path acquire that observes it can
+        // dispatch without re-checking the pool. The s3_wrapper backend handle is a process-wide static, hence
+        // the single-plugin lock.
+        _plugin = plugin;
+        // per-worker pool: each thread owns an ObjectStorageWorker (from the factory) with its own in-flight window
+        _object_storage_pool = std::make_unique<utils::ThreadPool<Workload>>(_object_storage_factory, _object_storage_size);
+        _ready_plugin.store(static_cast<int>(plugin), std::memory_order_release);
+        return common::ResponseCode::Success;
```
[the remark this answers is not in the document]
Comment (noa-neria) [reply cpp/streamer/impl/pools/backend_pools.cc:73]: **Valid — fixed by making the first-submission branch strong-exception-safe (commit the lock only after the pool exists), which is a bit cleaner than retrying in the already-locked branch.**

You're right that `ThreadPool`'s constructor spawns worker threads and can throw `std::system_error` under a thread/fd/process `ulimit` (not just OOM), and the old order committed `_plugin` before the pool:

    _plugin = plugin;                                              // committed first
    _object_storage_pool = std::make_unique<ThreadPool<...>>(...); // could throw -> pool stays null

...so a retry saw the plugin locked with a null pool and `_pools.push` would `ASSERT`-abort.

Now the pool is built first, and `_plugin` / `_ready_plugin` are committed only after it succeeds:

    _object_storage_pool = std::make_unique<ThreadPool<Workload>>(...);  // throws before any commit
    _plugin = plugin;
    _ready_plugin.store(static_cast<int>(plugin), std::memory_order_release);

If construction throws, nothing is committed — `_plugin` stays empty, `_ready_plugin` stays `-1`, pool stays null — so the next same-plugin submission re-enters this branch and retries cleanly. The slow path is serialized by `_plugin_mutex`, so no concurrent submitter can observe the half-built state, and the fast path still gates on `_ready_plugin` (published last,

**Review by omer-dayan — APPROVED** [review]

**Comment by noa-neria** [comment]
Re: `cpp/mock/streamer-mock.cc` — mock returns -1 (not `FINISHED_ERROR_CODE=1`) on drain, diverging from the PERSISTENT-responder contract

Acknowledged, and intentionally deferred to the concurrent-submission (ring-buffer) next branch — it's a test-fidelity gap with no production or current-test impact.

Why it doesn't bite today: every test that drives the mock is single-submission and count-bounded — the caller pulls exactly the number of responses it submitted and never issues a read past the last one. So the drained path this comment is about is never exercised, in tests or in the real callers (vLLM/SGLang), which also stop at their known response count.

One correction to the suggested fix: the target value on drain isn't `FINISHED_ERROR_CODE` either. In PERSISTENT mode the real responder, when drained-but-still-open, blocks (timeout 0) or returns `TimedOut` (timeout > 0); `FinishedError` is reserved for teardown (runai_end / stop), not for "you've drained the current submissions." So swapping -1 → FINISHED_ERROR_CODE would trade one wrong contract for another.

Plan: the mock's faithful fix is coupled to the ring-buffer work, where it needs a larger rework anyway — per-submission state + counters (today a second submission clobbers the first's), honoring timeout_ms, and the correct drain behavior (block / return TimedOut, FinishedError only on teardown). Doing it now, ahead of multi-submission tests that would actually verify it, would be unverified churn. It's tracked in ring-buffer.md with this exact contract.

---
Note: the other two caveats from the review are already handled in this branch —
- list_files handle_object_store-before-try → fixed (moved inside the try, temp streamer now runai_end'd on a credential-resolution throw).
- discard() no-op default trap → fixed (made pure virtual; the two test workers now carry explicit empty overrides).``

File: cpp/s3/s3_mock/s3_mock.cc [hunk 12]
```diff
@@ -450,6 +473,17 @@ void obj_free_file_list(common::backend_api::ObjectFileEntry_t* entries, unsigne
...
 
+const char* runai_mock_s3_last_client_config_value(const char* key)
+{
+    const auto guard = std::unique_lock<std::mutex>(__mutex);
+    if (key == nullptr)
+    {
+        return nullptr;
+    }
+    auto it = __mock_last_client_config.find(key);
+    return (it != __mock_last_client_config.end()) ? it->second.c_str() : nullptr;
```
[the remark this answers is not in the document]
Comment (noa-neria) [reply cpp/s3/s3_mock/s3_mock.cc]: Fixed
